src/libopensc/card-piv.c

```diff
@@ -509,212 +842,1990 @@ static struct sc_card_driver piv_drv = {
+		r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+		goto err;
+	}
+#else
+	if(!EVP_MAC_init(cmac_ctx, (const unsigned char *)priv->sm_session.SKrmac,
+				priv->sm_session.aes_size, cmac_params)
+			|| !EVP_MAC_update(cmac_ctx, priv->sm_session.R_MCV, MCVlen)
+			|| !EVP_MAC_update(cmac_ctx, sm_apdu->resp, macdatalen)
+			|| !EVP_MAC_final(cmac_ctx, priv->sm_session.R_MCV, &R_MCVlen, MCVlen)) {
+		piv_log_openssl(card->ctx);
+		r = SC_ERROR_INTERNAL;
+		goto err;
+	}
+#endif
+
+	if (memcmp(priv->sm_session.R_MCV, rmac8.value, 8) != 0) {
+		sc_log(card->ctx, "SM 8 bytes of R-MAC do not match received R-MAC");
+		r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+		goto err;
+	}
+
+	ed_ctx = EVP_CIPHER_CTX_new();
+	if (ed_ctx == NULL) {
+		r = SC_ERROR_INTERNAL;
+		goto err;
+	}
+
+	/* generate same IV used to encrypt response on card */
+	if (EVP_EncryptInit_ex(ed_ctx, (*cs->cipher_ecb)(), NULL, priv->sm_session.SKenc, zeros) != 1
+			|| EVP_CIPHER_CTX_set_padding(ed_ctx,0) != 1
+			|| EVP_EncryptUpdate(ed_ctx, IV, &outli, priv->sm_session.resp_enc_counter, 16) != 1
+			|| EVP_EncryptFinal_ex(ed_ctx, discard, &outdl) != 1
+			|| outdl != 0) {
+		sc_log(card->ctx,"SM encode failed in OpenSSL");
+		piv_log_openssl(card->ctx);
+		r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+		goto err;
+	}
+
+	/* some commands do not have response data */
+	if (ee.value != NULL) {
+		p = ee.value;
+		inlen = ee.len;
+		if (inlen < 17 || *p != 0x01) { /*padding indicator is required */
+			sc_log(card->ctx, "SM padding indicator not 0x01");
+			r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+			goto err;
+		}
+
+		p++; /* skip padding indicator */
+		inlen --;
+
+		if ((inlen % 16) != 0) {
+			sc_log(card->ctx,"SM encrypted data not multiple of 16");
+			r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+			goto err;
+		}
+
+		/*
+		 * Encrypted data has 1 to 16 pad bytes, so may be 1 to 16 bytes longer
+		 * then expected. i.e. plain->resp and resplen.So will do last block
+		 * and recombine.
+		 */
+
+		inlen -= 16;
+		if (plain->resplen < (unsigned) inlen || plain->resp == NULL) {
+			sc_log(card->ctx, "SM response will not fit in resp,resplen");
+			r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+			goto err;
+		}
+
+		q = plain->resp;
+
+		/* first round encryptes counter with zero IV, and does not save the output */
+		if (EVP_CIPHER_CTX_reset(ed_ctx) != 1
+				|| EVP_DecryptInit_ex(ed_ctx, (*cs->cipher_cbc)(), NULL, priv->sm_session.SKenc, IV) != 1
+				|| EVP_CIPHER_CTX_set_padding(ed_ctx,0) != 1
+				|| EVP_DecryptUpdate(ed_ctx, q ,&outl, p, inlen) != 1
+				|| EVP_DecryptUpdate(ed_ctx, lastb, &outll, p + inlen, 16 ) != 1
+				|| EVP_DecryptFinal_ex(ed_ctx, discard, &outdl) != 1
+				|| outdl != 0
+				|| outll != 16) {  /* should not happen */
+			sc_log(card->ctx,"SM _decode failed in OpenSSL");
+			piv_log_openssl(card->ctx);
+			r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+			goto err;
+		}
+
+		/* unpad last block and get bytes in last block */
+		for (i = 15; i >=  0 ; i--) {
```

Comment:
Is the condition off by one? If the last block had only zero bytes, the loop would exit with `i == -1`, which is used as an array index immediately after the loop.

Reply:
The last block can not have zero bytes. 

sp800-73-4 Part2 "4.2.2 Command and Response Data Confidentiality" says:
"Padding: Prior to encryption, one to sixteen bytes of padding data shall be appended to the PIV data. The padding shall be '80' followed by the number of zeros needed to make the total length of the message to be encrypted (PIV data plus padding) a multiple of sixteen bytes. The first byte of the value field of tag '87', the padding-content indicator byte, shall be '01' to indicate that padding has been applied."

In line 1378, the pointer `lastb` and size `outll` of last block are returned. `outll` is then checked in line 1381 that it is 16 bytes: https://github.com/OpenSC/OpenSC/pull/2053/files/6135b605e90a7dda2faf5c633b16c2fadc01c88b#diff-14d802e467876e00ece49c8d6621ec74eaf7d02384642ac0694de83397d9bbfcR1378-R1381

So the loop starts at the last byte which should be `00` or `80` and looks back till it finds the `80` which is the start of the padding to get the length of the message without padding. There must always be padding. The decryption is considered invalid if the padding is not found at end. The padding is 1 to 16 bytes.  

So I think the code is correct. 

Also look at  lines 1350, 1352, 1364 which make sure decrypted message is multiple of 16 and fetch the last 16 bytes in listb.

Reply:
The check is sufficient for correct implementations. My concern was that a defective implementation could trigger an out-of-bounds read by returning a message that has 16 zeros in the end when decrypted.

Reply:
OK how about:

diff --git a/src/libopensc/card-piv.c b/src/libopensc/card-piv.c
index 85a646b61..4f51bc0e9 100644
--- a/src/libopensc/card-piv.c
+++ b/src/libopensc/card-piv.c
@@ -1396,7 +1396,7 @@ static int piv_decode_apdu(sc_card_t *card, sc_apdu_t *plain, sc_apdu_t *sm_apdu
                        goto err;
                }
 
-               if (lastb[i] != 0x80) {
+               if (i < 0 || lastb[i] != 0x80) {
                        sc_log(card->ctx, "SM Padding not correct");
                        r = SC_ERROR_SM_AUTHENTICATION_FAILED;
                        goto err;

Reply:
OK, pushed a fix.